presto-main/src/main/java/com/facebook/presto/operator/aggregation/AccumulatorCompiler.java

```diff
@@ -553,33 +596,69 @@ private static void generateAddIntermediateAsCombine(
         BytecodeBlock body = method.getBody();
         Variable thisVariable = method.getThis();
 
-        Variable block = scope.getVariable("block");
-        Variable scratchState = scope.declareVariable(singleStateClass, "scratchState");
+        int stateCount = stateFieldAndMetadatas.size();
+        List<Variable> scratchState = new ArrayList<>();
+        for (int i = 0; i < stateCount; i++) {
+            Class<?> scratchStateClass = stateFieldAndMetadatas.get(i).getStateInfo().getFactory().getSingleStateClass();
+            scratchState.add(scope.declareVariable(scratchStateClass, "scratchState_" + i));
+        }
+
+        List<Variable> block;
+        if (stateCount == 1) {
+            block = ImmutableList.of(scope.getVariable("block"));
+        }
+        else {
+            Variable columnarRow = scope.declareVariable(ColumnarRow.class, "columnarRow");
+            body.append(columnarRow.set(
```

Comment:
I don't think you need ColumnarRow here. Let's talk about the high levels of this class and how you intended to change it.

Reply:
@haozhun Yeah. I am using `RowBlock` as a `Page` here.